File: elixir/test/contrato_test.exs
defmodule Rinha.ContratoTest do
  @moduledoc """
  O contrato HTTP do doc `02-regras.md`, endpoint a endpoint.

  Espelha `fastapi/tests/test_contrato.py`. Um caso que existe lá e não existe
  aqui seria uma diferença de cobertura entre projetos, e cobertura diferente
  invalida a comparação tanto quanto código diferente.
  """

  use Rinha.Caso, async: false

  describe "POST /clientes/:id/transacoes" do
    test "crédito devolve 200 com limite e saldo atualizados" do
      conn = transacionar(1, %{valor: 1000, tipo: "c", descricao: "salario"})

      assert conn.status == 200
      assert Jason.decode!(conn.resp_body) == %{"limite" => 100_000, "saldo" => 1000}
    end

    test "débito dentro do limite devolve 200 com saldo negativo" do
      conn = transacionar(1, %{valor: 1000, tipo: "d", descricao: "aluguel"})

      assert conn.status == 200
      assert Jason.decode!(conn.resp_body) == %{"limite" => 100_000, "saldo" => -1000}
    end

    test "débito que estoura o limite devolve 422 e NÃO altera o saldo" do
      conn = transacionar(1, %{valor: 100_001, tipo: "d", descricao: "caro"})

      assert conn.status == 422
      # A parte que importa: a recusa é atômica. Um 422 que tivesse debitado
      # metade seria pior que um 500.
      assert saldo_de(1) == 0
    end

    test "débito exatamente no limite é aceito" do
      conn = transacionar(1, %{valor: 100_000, tipo: "d", descricao: "limite"})

      assert conn.status == 200
      assert saldo_de(1) == -100_000
    end

    test "cliente inexistente devolve 404" do
      assert transacionar(6, %{valor: 1, tipo: "c", descricao: "x"}).status == 404
    end

    test "id não numérico devolve 404" do
      assert requisitar(:post, "/clientes/abc/transacoes", "{}").status == 404
    end

    test "payloads inválidos devolvem 422" do
      invalidos = [
        %{valor: 1.2, tipo: "c", descricao: "float"},
        %{valor: 0, tipo: "c", descricao: "zero"},
        %{valor: -5, tipo: "c", descricao: "negativo"},
        %{valor: true, tipo: "c", descricao: "booleano"},
        %{valor: "100", tipo: "c", descricao: "string"},
        %{valor: 100, tipo: "x", descricao: "tipo"},
        %{valor: 100, tipo: "c", descricao: "onze caracteres"},
        %{valor: 100, tipo: "c", descricao: ""},
        %{valor: 100, tipo: "c"},
        %{valor: 100, tipo: "c", descricao: nil}
      ]

      for payload <- invalidos do
        conn = transacionar(1, payload)
        assert conn.status == 422, "esperava 422 para #{inspect(payload)}"
      end
    end

    test "corpo que não é JSON devolve 422" do
      assert requisitar(:post, "/clientes/1/transacoes", "não é json").status == 422
    end

    test "descrição de 10 caracteres é aceita e a de 11 não" do
      assert transacionar(1, %{valor: 1, tipo: "c", descricao: "1234567890"}).status == 200
      assert transacionar(1, %{valor: 1, tipo: "c", descricao: "12345678901"}).status == 422
    end
  end

  describe "GET /clientes/:id/extrato" do
    test "cliente sem transações devolve saldo zero e lista vazia" do
      conn = requisitar(:get, "/clientes/1/extrato")
      corpo = Jason.decode!(conn.resp_body)

      assert conn.status == 200
      assert corpo["saldo"]["total"] == 0
      assert corpo["saldo"]["limite"] == 100_000
      assert corpo["ultimas_transacoes"] == []
      assert String.ends_with?(corpo["saldo"]["data_extrato"], "Z")
    end

    test "devolve no máximo 10 transações, da mais recente para a mais antiga" do
      for i <- 1..15 do
        transacionar(1, %{valor: i, tipo: "c", descricao: "t#{i}"})
      end

      corpo = Jason.decode!(requisitar(:get, "/clientes/1/extrato").resp_body)

      assert length(corpo["ultimas_transacoes"]) == 10
      # A 15ª é a primeira da lista; a 6ª é a última. As cinco primeiras ficaram
      # de fora.
      assert Enum.map(corpo["ultimas_transacoes"], & &1["valor"]) == Enum.to_list(15..6//-1)
    end

    test "read-your-writes: o extrato logo após o POST já enxerga a transação" do
      transacionar(1, %{valor: 777, tipo: "c", descricao: "agora"})
      corpo = Jason.decode!(requisitar(:get, "/clientes/1/extrato").resp_body)

      assert corpo["saldo"]["total"] == 777
      assert hd(corpo["ultimas_transacoes"])["valor"] == 777
    end

    test "cliente inexistente devolve 404" do
      assert requisitar(:get, "/clientes/6/extrato").status == 404
    end

    test "o campo realizada_em sai no formato do README" do
      transacionar(1, %{valor: 1, tipo: "c", descricao: "fmt"})
      corpo = Jason.decode!(requisitar(:get, "/clientes/1/extrato").resp_body)

      assert hd(corpo["ultimas_transacoes"])["realizada_em"] =~
               ~r/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{6}Z$/
    end
  end

  describe "content-type" do
    test "é application/json puro, sem charset" do
      conn = requisitar(:get, "/clientes/1/extrato")
      assert Plug.Conn.get_resp_header(conn, "content-type") == ["application/json"]
    end
  end
end

File: go/contrato_test.go
// O contrato HTTP da Rinha, verificado endpoint a endpoint.
//
// Porte de `elixir/test/contrato_test.exs` e `fastapi/tests/test_contrato.py`.
// Cada asserção aqui corresponde a uma verificação da simulação oficial —
// `.claude/docs/02-regras.md`, seções 1 e 6.
package main

import (
	"net/http"
	"testing"
)

func TestExtratoDeClienteExistente(t *testing.T) {
	h := caso(t)

	resposta := extratoDe(h, 1)
	if resposta.Code != http.StatusOK {
		t.Fatalf("status %d, esperado 200", resposta.Code)
	}

	corpo := decodificar(t, resposta.Body.Bytes())
	saldo := corpo["saldo"].(map[string]any)
	if saldo["limite"].(float64) != 100_000 {
		t.Errorf("limite %v, esperado 100000", saldo["limite"])
	}
	if saldo["total"].(float64) != 0 {
		t.Errorf("total %v, esperado 0", saldo["total"])
	}
	if saldo["data_extrato"] == "" {
		t.Error("data_extrato vazia")
	}
	// Lista vazia, e não `null`: o contrato mostra um array, e um `null` faria o
	// Gatling falhar ao ler `ultimas_transacoes.size()`.
	if ultimas, ok := corpo["ultimas_transacoes"].([]any); !ok || len(ultimas) != 0 {
		t.Errorf("ultimas_transacoes %v, esperado lista vazia", corpo["ultimas_transacoes"])
	}
}

func TestClienteInexistenteDa404(t *testing.T) {
	h := caso(t)

	// O cliente 6 NÃO existe, e o teste oficial verifica exatamente isso.
	if resposta := extratoDe(h, 6); resposta.Code != http.StatusNotFound {
		t.Errorf("GET extrato do 6: status %d, esperado 404", resposta.Code)
	}
	resposta := transacionar(h, 6, `{"valor":1,"tipo":"c","descricao":"x"}`)
	if resposta.Code != http.StatusNotFound {
		t.Errorf("POST no 6: status %d, esperado 404", resposta.Code)
	}
	// Um ID que nem número é: tem de ser 404, não 500.
	if r := requisitar(h, http.MethodGet, "/clientes/abc/extrato", ""); r.Code != http.StatusNotFound {
		t.Errorf("GET /clientes/abc/extrato: status %d, esperado 404", r.Code)
	}
}

func TestTransacaoDeSucessoResponde200(t *testing.T) {
	h := caso(t)

	// 200, obrigatoriamente — nunca 201. É uma das regras explícitas do README.
	resposta := transacionar(h, 1, `{"valor":1000,"tipo":"c","descricao":"toma"}`)
	if resposta.Code != http.StatusOK {
		t.Fatalf("status %d, esperado 200", resposta.Code)
	}
	if tipo := resposta.Header().Get("Content-Type"); tipo != tipoJSON {
		t.Errorf("content-type %q, esperado %q", tipo, tipoJSON)
	}

	corpo := decodificar(t, resposta.Body.Bytes())
	if corpo["limite"].(float64) != 100_000 || corpo["saldo"].(float64) != 1000 {
		t.Errorf("corpo %v, esperado limite 100000 e saldo 1000", corpo)
	}
}

func TestDebitoDentroDoLimitePassa(t *testing.T) {
	h := caso(t)

	resposta := transacionar(h, 1, `{"valor":100000,"tipo":"d","descricao":"limite"}`)
	if resposta.Code != http.StatusOK {
		t.Fatalf("status %d, esperado 200", resposta.Code)
	}
	// Exatamente no limite: saldo -limite é válido, -limite-1 não.
	if saldo := saldoDe(t, 1); saldo != -100_000 {
		t.Errorf("saldo %d, esperado -100000", saldo)
	}
}

func TestDebitoQueEstouraOLimiteDa422ENaoAplica(t *testing.T) {
	h := caso(t)

	resposta := transacionar(h, 1, `{"valor":100001,"tipo":"d","descricao":"estoura"}`)
	if resposta.Code != http.StatusUnprocessableEntity {
		t.Fatalf("status %d, esperado 422", resposta.Code)
	}
	// "sem aplicar a transação": nem o saldo muda, nem sobra transação no
	// extrato. É o que a transação de banco garante.
	if saldo := saldoDe(t, 1); saldo != 0 {
		t.Errorf("saldo %d, esperado 0", saldo)
	}
	corpo := decodificar(t, extratoDe(h, 1).Body.Bytes())
	if ultimas := corpo["ultimas_transacoes"].([]any); len(ultimas) != 0 {
		t.Errorf("%d transações no extrato, esperado nenhuma", len(ultimas))
	}
}

func TestPayloadsInvalidosDao422(t *testing.T) {
	h := caso(t)

	// Os cinco payloads que a simulação oficial manda de propósito, mais os que
	// as outras três stacks já cobriam.
	casos := map[string]string{
		"valor fracionário":  `{"valor":1.2,"tipo":"c","descricao":"x"}`,
		"valor booleano":     `{"valor":true,"tipo":"c","descricao":"x"}`,
		"valor negativo":     `{"valor":-1,"tipo":"c","descricao":"x"}`,
		"valor zero":         `{"valor":0,"tipo":"c","descricao":"x"}`,
		"valor como string":  `{"valor":"1","tipo":"c","descricao":"x"}`,
		"tipo inválido":      `{"valor":1,"tipo":"x","descricao":"x"}`,
		"descrição com 11":   `{"valor":1,"tipo":"c","descricao":"123456789 e mais um pouco"}`,
		"descrição vazia":    `{"valor":1,"tipo":"c","descricao":""}`,
		"descrição null":     `{"valor":1,"tipo":"c","descricao":null}`,
		"descrição ausente":  `{"valor":1,"tipo":"c"}`,
		"valor ausente":      `{"tipo":"c","descricao":"x"}`,
		"corpo vazio":        ``,
		"corpo não é objeto": `[]`,
		"JSON malformado":    `{"valor":`,
	}

	for nome, corpo := range casos {
		t.Run(nome, func(t *testing.T) {
			resposta := transacionar(h, 1, corpo)
			if resposta.Code != http.StatusUnprocessableEntity {
				t.Errorf("status %d, esperado 422", resposta.Code)
			}
		})
	}

	// Nenhum deles pode ter tocado no saldo.
	if saldo := saldoDe(t, 1); saldo != 0 {
		t.Errorf("saldo %d depois dos payloads inválidos, esperado 0", saldo)
	}
}

func TestDescricaoComExatamente10CaracteresPassa(t *testing.T) {
	h := caso(t)

	// A fronteira: 10 é válido, 11 não. Um `<` no lugar de `<=` passaria em
	// todos os outros testes.
	if r := transacionar(h, 1, `{"valor":1,"tipo":"c","descricao":"1234567890"}`); r.Code != http.StatusOK {
		t.Errorf("10 caracteres: status %d, esperado 200", r.Code)
	}
	if r := transacionar(h, 1, `{"valor":1,"tipo":"c","descricao":"12345678901"}`); r.Code != http.StatusUnprocessableEntity {
		t.Errorf("11 caracteres: status %d, esperado 422", r.Code)
	}
	// Acentos contam como UM caractere, e não como os dois bytes do UTF-8:
	// `len()` sobre bytes recusaria esta descrição de 10 letras.
	if r := transacionar(h, 1, `{"valor":1,"tipo":"c","descricao":"ãéíõúçãéí"}`); r.Code != http.StatusOK {
		t.Errorf("9 caracteres acentuados: status %d, esperado 200", r.Code)
	}
}

func TestExtratoEmOrdemDecrescente(t *testing.T) {
	h := caso(t)

	// A sequência exata da fase 2 do Gatling: crédito "toma", depois débito
	// "devolve" — e o extrato tem de mostrar "devolve" primeiro.
	transacionar(h, 1, `{"valor":1,"tipo":"c","descricao":"toma"}`)
	transacionar(h, 1, `{"valor":1,"tipo":"d","descricao":"devolve"}`)

	corpo := decodificar(t, extratoDe(h, 1).Body.Bytes())
	ultimas := corpo["ultimas_transacoes"].([]any)
	if len(ultimas) != 2 {
		t.Fatalf("%d transações, esperado 2", len(ultimas))
	}
	primeira := ultimas[0].(map[string]any)
	segunda := ultimas[1].(map[string]any)
	if primeira["descricao"] != "devolve" || primeira["tipo"] != "d" {
		t.Errorf("primeira %v, esperado devolve/d", primeira)
	}
	if segunda["descricao"] != "toma" || segunda["tipo"] != "c" {
		t.Errorf("segunda %v, esperado toma/c", segunda)
	}
}

func TestExtratoLimitaEm10Transacoes(t *testing.T) {
	h := caso(t)

	for range 15 {
		transacionar(h, 1, `{"valor":1,"tipo":"c","descricao":"x"}`)
	}

	corpo := decodificar(t, extratoDe(h, 1).Body.Bytes())
	if ultimas := corpo["ultimas_transacoes"].([]any); len(ultimas) != qtdTransacoesExtrato {
		t.Errorf("%d transações, esperado %d", len(ultimas), qtdTransacoesExtrato)
	}
	// `total` é o saldo TOTAL, não a soma das 10 listadas.
	saldo := corpo["saldo"].(map[string]any)
	if saldo["total"].(float64) != 15 {
		t.Errorf("total %v, esperado 15 (todas as transações, não só as 10 listadas)", saldo["total"])
	}
}

func TestReadYourWrites(t *testing.T) {
	h := caso(t)

	// Fase 2 do Gatling: um POST seguido de 5 GETs paralelos, todos exigindo ver
	// a transação e o saldo exato que o POST devolveu.
	resposta := transacionar(h, 1, `{"valor":777,"tipo":"c","descricao":"danada"}`)
	saldoDoPost := decodificar(t, resposta.Body.Bytes())["saldo"].(float64)

	for range 5 {
		corpo := decodificar(t, extratoDe(h, 1).Body.Bytes())
		if corpo["saldo"].(map[string]any)["total"].(float64) != saldoDoPost {
			t.Fatalf("extrato não vê o saldo que o POST devolveu (%v)", saldoDoPost)
		}
		ultimas := corpo["ultimas_transacoes"].([]any)
		if len(ultimas) == 0 || ultimas[0].(map[string]any)["descricao"] != "danada" {
			t.Fatal("extrato não vê a transação que o POST acabou de gravar")
		}
	}
}

func TestMetodoErradoNaRotaCerta(t *testing.T) {
	h := caso(t)

	// `GET` em /transacoes não casa com `POST /clientes/{id}/transacoes` e cai
	// no handler curinga. 404 é o que as outras stacks respondem.
	if r := requisitar(h, http.MethodGet, "/clientes/1/transacoes", ""); r.Code != http.StatusNotFound {
		t.Errorf("status %d, esperado 404", r.Code)
	}
	if r := requisitar(h, http.MethodGet, "/", ""); r.Code != http.StatusNotFound {
		t.Errorf("status %d, esperado 404", r.Code)
	}
}
